src/types.ts

```diff
@@ -4533,19 +4557,65 @@ export interface LevelOpt {
 	 * Called when encountered a symbol not defined in "tiles".
 	 */
 	wildcardTile?: (sym: string, pos: Vec2) => CompList<any> | null | undefined,
+	/**
+	 * Enables pathfinding if true.
+	 */
+	usePathfinding?: boolean
 }
 
 export interface TileComp extends Comp {
 	tilePos: Vec2,
 	setTilePos(...args),
+	getLevel(): GameObj<any>,
 	moveLeft(),
 	moveRight(),
 	moveUp(),
 	moveDown(),
 }
 
 export interface ObstacleComp extends Comp {
-	isObstacle: boolean,
+	set isObstacle(isObstacle: boolean)
+	get isObstacle(): boolean,
+}
+
+export interface CostComp extends Comp {
+    cost: number
+}
+
+export enum EdgeMask {
+    None = 0,
+    Left = 1,
+    Top = 2,
+    LeftTop = 3,
+    Right = 4,
+    Horizontal = 5,
+    RightTop = 6,
+    HorizontalTop = 7,
+    Bottom = 8,
+    LeftBottom = 9,
+    Vertical = 10,
+    LeftVertical = 11,
+    RightBottom = 12,
+    HorizontalBottom = 13,
+    RightVertical = 14,
+    All = 15,
+}
+
+export interface EdgesComp extends Comp {
+    mask: EdgeMask
+}
+
+export interface PathfindingComp extends Comp {
+	invalidateNavigationMap(),
+	onNavigationMapChanged(cb: () => void): EventController,
```

Comment:
When would a end user use `invalidateNavigationMap()` manually, or listen to `onNavigationMapChanged()`?

Reply:
Agent listens to it, because it needs to know when its path became invalid and needs to be recomputed. A user building a bot might need to know it too.
invalidateNavigationMap is not private since theoretically you can make changes to the level which can't be detected by us.

Reply:
What kind of changes we can't detect?

Reply:
A user can remove the obstacle component for example. Or Add an isObstacle: true without using obstacle(). Or even a get isObstacle() in case the object determines the obstacle state from other states.